state/execution.go

```diff
@@ -218,26 +224,32 @@ func (s *State) validateBlock(block *types.Block) error {
 	return nil
 }
 
-//-----------------------------------------------------------------------------
-// ApplyBlock executes the block, then commits and updates the mempool atomically
-
-// Execute and commit block against app, save block and state
+// ApplyBlock executes the block, then commits and updates the mempool
+// atomically, optionally indexing transaction results.
 func (s *State) ApplyBlock(eventCache types.Fireable, proxyAppConn proxy.AppConnConsensus,
 	block *types.Block, partsHeader types.PartSetHeader, mempool types.Mempool) error {
 
-	// Run the block on the State:
-	// + update validator sets
-	// + run txs on the proxyAppConn
-	err := s.ExecBlock(eventCache, proxyAppConn, block, partsHeader)
+	txResults, err := s.ExecBlock(eventCache, proxyAppConn, block, partsHeader)
 	if err != nil {
-		return errors.New(Fmt("Exec failed for application: %v", err))
+		return fmt.Errorf("Exec failed for application: %v", err)
 	}
 
 	// lock mempool, commit state, update mempoool
 	err = s.CommitStateUpdateMempool(proxyAppConn, block, mempool)
 	if err != nil {
-		return errors.New(Fmt("Commit failed for application: %v", err))
+		return fmt.Errorf("Commit failed for application: %v", err)
 	}
+
+	batch := txindexer.NewBatch()
+	for i, r := range txResults {
+		if r != nil {
```

Comment:
is this ever nil, now that we assume the ABCI return vals are in order and complete?